addon/-private/properties/collection/main.js

```diff
@@ -0,0 +1,104 @@
+/* global Symbol */
+import Ember from 'ember';
+import { buildSelector, assign } from '../../helpers';
+import { create } from '../../create';
+import { count } from '../count';
+import Ceibo from 'ceibo';
+
+export class Collection {
+  constructor(definition, parent, key) {
+    this.definition = definition;
+    this.parent = parent;
+    this.key = key;
+
+    this._itemCounter = create({
+      count: count(definition.scope, {
+        resetScope: definition.resetScope,
+        testContainer: definition.testContainer
+      })
+    }, { parent });
+
+    this._items = [];
+  }
+
+  get length() {
+    return this._itemCounter.count;
+  }
+
+  objectAt(index) {
+    if (typeof this._items[index] === 'undefined') {
+      let { definition, parent, key } = this;
+      let scope = buildSelector({}, definition.scope, { at: index });
+
+      let finalizedDefinition = assign({}, definition);
+
+      finalizedDefinition.scope = scope;
+
+      let tree = create(finalizedDefinition, { parent });
+
+      // Change the key of the root node
+      Ceibo.meta(tree).key = `${key}[${index}]`;
+
+      this._items[index] = tree;
+    }
+
+    return this._items[index];
+  }
+
+  filter(...args) {
+    return this.toArray().filter(...args);
+  }
+
+  filterBy(...args) {
+    return this.toArray().filterBy(...args);
+  }
+
+  forEach(...args) {
+    return this.toArray().forEach(...args);
+  }
+
+  map(...args) {
+    return this.toArray().map(...args);
+  }
+
+  mapBy(...args) {
+    return this.toArray().mapBy(...args);
+  }
+
+  toArray() {
+    let { length } = this;
+
+    let array = Ember.A();
+
+    for (let i = 0; i < length; i++) {
+      array.push(this.objectAt(i));
+    }
+
+    return array;
+  }
```

Comment:
Just an idea...
```javascript
toArray() {
  return Ember.A(Array.from(this));
}

...

if (typeof (Symbol) !== 'undefined' && Symbol.iterator) {
  Collection.prototype[Symbol.iterator] = function() {
    let i = 0;
    let next = () => ({ done: i >= this.length, value: this.objectAt(i++) });
    return { next };
  }
}
```

Reply:
that would be cool, but then we would also be assume that `Symbol` and `Symbol.iterator` are supported no?

Reply:
Ah, yeah...good point.